### ISSUE TITLE
Problems running lerna on Alpine?

### ISSUE THREAD

**Issue by FezVrasta** [lerna/lerna#752]
## Expected Behavior

Trying to run `lerna bootstrap` on Docker on `mhart/alpine-node:7.5` image with [this repro](https://github.com/FezVrasta/docker-yarn-lerna-bug) everything should work fine.

## Current Behavior

I get:

```
$ lerna bootstrap
Lerna v2.0.0-rc.1
Independent Versioning Mode
Bootstrapping 8 packages
Preinstalling packages
Installing external dependencies
NpmUtilities.installInDir     ("./packages/js-utils", ["babel-cli@^6.22.2","babel-core@^6.22.1","babel-preset-es2015@^6.22.0","jest...)
Command exited with status 1: yarn install
Errored while running BootstrapCommand.execute
Command exited with status 1: yarn install
Waiting for 3 child processes to exit. CTRL-C to exit immediately.
```

Instead, on macOS everything works fine.

## Steps to Reproduce (for bugs)
1. Use this repository: https://github.com/FezVrasta/docker-yarn-lerna-bug
2. Build the container

<details><summary>lerna.json</summary><p>

```json
{
  "lerna": "2.0.0-rc.1",
  "packages": [
    "packages/*"
  ],
  "version": "independent",
  "npmClient": "yarn"
}

```
</p></details>

<details><summary>lerna-debug.log</summary><p>

```txt
bash-4.3# cat lerna-debug.log
lerna(verbose) GitUtilities.isInitialized    ()
lerna(verbose) GitUtilities.isInitialized    () => true
lerna(verbose) GitUtilities.getTopLevelDirectory()
lerna(verbose) GitUtilities.getTopLevelDirectory() => "."
lerna(verbose) FileSystemUtilities.existsSync("./lerna.json")
lerna(verbose) FileSystemUtilities.existsSync("./lerna.json") => true
lerna(verbose) FileSystemUtilities.readFileSync("./lerna.json")
lerna(verbose) FileSystemUtilities.readFileSync("./lerna.json") => "{\n  \"lerna\": \"2.0.0-rc.1\",\n  \"packages\": [\n    \"packages/*\"\n  ],\n  \"version\": \"independent\",\n  \"npmClient\": \"yarn\"\n}"
lerna(verbose) FileSystemUtilities.existsSync("./package.json")
lerna(verbose) FileSystemUtilities.existsSync("./package.json") => true
lerna(verbose) FileSystemUtilities.readFileSync("./package.json")
lerna(verbose) FileSystemUtilities.readFileSync("./package.json") => "{\n  \"devDependencies\": {\n    \"lerna\": \"2.0.0-rc.1\"\n  },\n  \"scripts\": {\n    \"postinstall\": \"lerna bootstrap\",\n    \"bootstrap\": \"lerna bootstrap\",\n    \"test-ci\": \"lerna run test-ci\"\n  }\n}"
lerna(info)    Lerna v2.0.0-rc.1
lerna(info)    Independent Versioning Mode
lerna(verbose) FileSystemUtilities.existsSync("./package.json")
lerna(verbose) FileSystemUtilities.existsSync("./package.json") => true
lerna(verbose) FileSystemUtilities.existsSync("./lerna.json")
lerna(verbose) FileSystemUtilities.existsSync("./lerna.json") => true
lerna(verbose) FileSystemUtilities.existsSync("./VERSION")
lerna(verbose) FileSystemUtilities.existsSync("./VERSION") => false
lerna(verbose) FileSystemUtilities.existsSync("./packages/eslint-config-****/package.json")
lerna(verbose) FileSystemUtilities.existsSync("./packages/eslint-config-****/package.json") => true
lerna(verbose) FileSystemUtilities.existsSync("./packages/js-events/package.json")
lerna(verbose) FileSystemUtilities.existsSync("./packages/js-events/package.json") => true
lerna(verbose) FileSystemUtilities.existsSync("./packages/js-utils/package.json")
lerna(verbose) FileSystemUtilities.existsSync("./packages/js-utils/package.json") => true
lerna(verbose) FileSystemUtilities.existsSync("./packages/postcss-color-function-custom-props/package.json")
lerna(verbose) FileSystemUtilities.existsSync("./packages/postcss-color-function-custom-props/package.json") => true
lerna(verbose) FileSystemUtilities.existsSync("./packages/postcss-what-input/package.json")
lerna(verbose) FileSystemUtilities.existsSync("./packages/postcss-what-input/package.json") => true
lerna(verbose) FileSystemUtilities.existsSync("./packages/****-styles/package.json")
lerna(verbose) FileSystemUtilities.existsSync("./packages/****-styles/package.json") => true
lerna(verbose) FileSystemUtilities.existsSync("./packages/react-components/package.json")
lerna(verbose) FileSystemUtilities.existsSync("./packages/react-components/package.json") => true
lerna(verbose) FileSystemUtilities.existsSync("./packages/webpack-postcss-config/package.json")
lerna(verbose) FileSystemUtilities.existsSync("./packages/webpack-postcss-config/package.json") => true
lerna(verbose) Attempting running BootstrapCommand.initialize
lerna(verbose) Successfully ran BootstrapCommand.initialize
lerna(verbose) Attempting running BootstrapCommand.execute
lerna(info)    Bootstrapping 8 packages
lerna(info)    Preinstalling packages
lerna(info)    Installing external dependencies
lerna(verbose) NpmUtilities.installInDir     ("./packages/js-events", ["babel-cli@^6.22.2","babel-core@^6.22.1","babel-preset-es2015@^6.22.0","bab...)
lerna(verbose) FileSystemUtilities.rename    ("./packages/js-events/package.json", "./packages/js-events/package.json.lerna_backup")
lerna(verbose) NpmUtilities.installInDir     ("./packages/js-utils", ["babel-cli@^6.22.2","babel-core@^6.22.1","babel-preset-es2015@^6.22.0","jest...)
lerna(verbose) FileSystemUtilities.rename    ("./packages/js-utils/package.json", "./packages/js-utils/package.json.lerna_backup")
lerna(verbose) NpmUtilities.installInDir     ("./packages/react-components", ["babel-core@6.17.0","babel-preset-es2015@^6.22.0","babel-preset-stag...)
lerna(verbose) FileSystemUtilities.rename    ("./packages/react-components/package.json", "./packages/react-components/package.json.lerna_backup")
lerna(verbose) NpmUtilities.installInDir     ("./packages/postcss-color-function-custom-props", ["color@^1.0.3","postcss@^5.2.15"], {"client":"yar...)
lerna(verbose) FileSystemUtilities.rename    ("./packages/postcss-color-function-custom-props/package.json", "./packages/postcss-color-function-cu...)
lerna(verbose) FileSystemUtilities.rename    ("./packages/react-components/package.json", "./packages/react-components/package.json.lerna_backup") =>
lerna(verbose) FileSystemUtilities.writeFile ("./packages/react-components/package.json", "{\"dependencies\":{\"babel-core\":\"6.17.0\",\"babel-pr...)
lerna(verbose) FileSystemUtilities.rename    ("./packages/js-events/package.json", "./packages/js-events/package.json.lerna_backup") =>
lerna(verbose) FileSystemUtilities.writeFile ("./packages/js-events/package.json", "{\"dependencies\":{\"babel-cli\":\"^6.22.2\",\"babel-core\":\"...)
lerna(verbose) FileSystemUtilities.rename    ("./packages/js-utils/package.json", "./packages/js-utils/package.json.lerna_backup") =>
lerna(verbose) FileSystemUtilities.writeFile ("./packages/js-utils/package.json", "{\"dependencies\":{\"babel-cli\":\"^6.22.2\",\"babel-core\":\"^...)
lerna(verbose) FileSystemUtilities.rename    ("./packages/postcss-color-function-custom-props/package.json", "./packages/postcss-color-function-cu...) =>
lerna(verbose) FileSystemUtilities.writeFile ("./packages/postcss-color-function-custom-props/package.json", "{\"dependencies\":{\"color\":\"^1.0....)
lerna(verbose) FileSystemUtilities.writeFile ("./packages/react-components/package.json", "{\"dependencies\":{\"babel-core\":\"6.17.0\",\"babel-pr...) =>
lerna(verbose) FileSystemUtilities.writeFile ("./packages/js-events/package.json", "{\"dependencies\":{\"babel-cli\":\"^6.22.2\",\"babel-core\":\"...) =>
lerna(verbose) FileSystemUtilities.writeFile ("./packages/js-utils/package.json", "{\"dependencies\":{\"babel-cli\":\"^6.22.2\",\"babel-core\":\"^...) =>
lerna(verbose) FileSystemUtilities.writeFile ("./packages/postcss-color-function-custom-props/package.json", "{\"dependencies\":{\"color\":\"^1.0....) =>
lerna(verbose) FileSystemUtilities.renameSync("./packages/postcss-color-function-custom-props/package.json.lerna_backup", "./packages/postcss-colo...)
lerna(verbose) FileSystemUtilities.renameSync("./packages/postcss-color-function-custom-props/package.json.lerna_backup", "./packages/postcss-colo...) =>
lerna(verbose) NpmUtilities.installInDir     ("./packages/postcss-color-function-custom-props", ["color@^1.0.3","postcss@^5.2.15"], {"client":"yar...) =>
lerna(verbose) NpmUtilities.installInDir     ("./packages/postcss-what-input", ["postcss@^5.2.15"], {"client":"yarn"})
lerna(verbose) FileSystemUtilities.rename    ("./packages/postcss-what-input/package.json", "./packages/postcss-what-input/package.json.lerna_back...)
lerna(verbose) FileSystemUtilities.rename    ("./packages/postcss-what-input/package.json", "./packages/postcss-what-input/package.json.lerna_back...) =>
lerna(verbose) FileSystemUtilities.writeFile ("./packages/postcss-what-input/package.json", "{\"dependencies\":{\"postcss\":\"^5.2.15\"}}")
lerna(verbose) FileSystemUtilities.writeFile ("./packages/postcss-what-input/package.json", "{\"dependencies\":{\"postcss\":\"^5.2.15\"}}") =>
lerna(verbose) FileSystemUtilities.renameSync("./packages/js-utils/package.json.lerna_backup", "./packages/js-utils/package.json")
lerna(verbose) FileSystemUtilities.renameSync("./packages/js-utils/package.json.lerna_backup", "./packages/js-utils/package.json") =>
lerna(error)   NpmUtilities.installInDir     ("./packages/js-utils", ["babel-cli@^6.22.2","babel-core@^6.22.1","babel-preset-es2015@^6.22.0","jest...)
    Command exited with status 1: yarn install
lerna(error)   Errored while running BootstrapCommand.execute
```
</p></details>

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 2.0.0-rc.1 |
| `npm --version`  | 4.1.2 |
| `yarn --version` | 0.22.0 |
| `node --version` | 7.5.0 |

| OS | Version |
| --- | --- |
| (Docker) Alpine Linux | 3.5.0 |

**Comment by FezVrasta** [comment]
I noticed that if I let lerna use `npm` instead of `yarn` everything works fine.

**Comment by FezVrasta** [comment]
Ok seems like the problem is when I run `lerna bootstrap` from `sh` instead of `bash`

**Comment by FezVrasta** [comment]
I found out this error inside `yarn-error.log` of one of the packages:

```
  Error: https://registry.yarnpkg.com/babel-cli/-/babel-cli-6.24.0.tgz: ENOENT: no such file or directory, open '/root/.cache/yarn/v1/npm-babel-cli-6.24.0-a05ffd210dca0c288a26d5319c5ac8669a265ad0/.yarn-tarball.tgz'
```

**Comment by FezVrasta** [comment]
I was able to create a repro to let you try it yourself:
https://github.com/FezVrasta/docker-yarn-lerna-bug

I opened an issue on the Yarn repository as well because I'm not sure who's fault...
https://github.com/yarnpkg/yarn/issues/3104

**Comment by evocateur** [comment]
Yeah, we definitely need to start using a network mutex when bootstrapping with `--npm-client=yarn` enabled. You could set concurrency to 1 in lerna.json targeting only the bootstrap command like this:

```json
{
  "lerna": "2.0.0-rc.1",
  "command": {
    "bootstrap": {
      "npmClient": "yarn",
      "concurrency": 1
    }
  },
  "version": "1.0.0"
}
```

(I namespaced the `npmClient` config under `bootstrap` because that's currently the only command that optionally uses a different npm client; all the other subcommands call `npm` regardless)

**Comment by evocateur** [comment]
A network mutex for `--npm-client=yarn` was added in #764, released in [v2.0.0-rc.3](https://github.com/lerna/lerna/releases/tag/v2.0.0-rc.3)
